synapse/storage/databases/main/deviceinbox.py

```diff
@@ -136,6 +137,79 @@ def process_replication_rows(self, stream_name, instance_name, token, rows):
     def get_to_device_stream_token(self):
         return self._device_inbox_id_gen.get_current_token()
 
+    async def get_new_messages(
+        self,
+        user_ids: Collection[str],
+        from_stream_id: int,
+        to_stream_id: int,
+    ) -> Dict[Tuple[str, str], List[JsonDict]]:
+        """
+        Retrieve to-device messages for a given set of user IDs.
+
+        Only to-device messages with stream ids between the given boundaries
+        (from < X <= to) are returned.
+
+        Note that a stream ID can be shared by multiple copies of the same message with
+        different recipient devices. Each (device, message_content) tuple has their own
+        row in the device_inbox table.
+
+        Args:
+            user_ids: The users to retrieve to-device messages for.
+            from_stream_id: The lower boundary of stream id to filter with (exclusive).
+            to_stream_id: The upper boundary of stream id to filter with (inclusive).
+
+        Returns:
+            A list of to-device messages.
+        """
+        # Bail out if none of these users have any messages
+        for user_id in user_ids:
+            if self._device_inbox_stream_cache.has_entity_changed(
+                user_id, from_stream_id
+            ):
+                break
+        else:
+            return {}
+
+        def get_new_messages_txn(txn: LoggingTransaction):
+            # Build a query to select messages from any of the given users that are between
+            # the given stream id bounds
+
+            # Scope to only the given users. We need to use this method as doing so is
+            # different across database engines.
+            many_clause_sql, many_clause_args = make_in_list_sql_clause(
+                self.database_engine, "user_id", user_ids
+            )
+
+            sql = f"""
+                SELECT user_id, device_id, message_json FROM device_inbox
+                WHERE {many_clause_sql}
+                AND ? < stream_id AND stream_id <= ?
+                ORDER BY stream_id ASC
+            """
+
+            txn.execute(sql, (*many_clause_args, from_stream_id, to_stream_id))
+
+            # Create a dictionary of (user ID, device ID) -> list of messages that
+            # that device is meant to receive.
+            recipient_user_id_device_id_to_messages: Dict[
+                Tuple[str, str], List[JsonDict]
+            ] = {}
+
+            for row in txn:
+                recipient_user_id = row[0]
+                recipient_device_id = row[1]
+                message_dict = db_to_json(row[2])
+
+                recipient_user_id_device_id_to_messages.setdefault(
+                    (recipient_user_id, recipient_device_id), []
+                ).append(message_dict)
+
+            return recipient_user_id_device_id_to_messages
+
+        return await self.db_pool.runInteraction(
+            "get_new_messages", get_new_messages_txn
+        )
+
```

Comment:
I'm rather uncomfortable that we're duplicating  `get_new_messages_for_device`. Can't we figure out the device ids and use `get_new_messages_for_device` ?

Reply:
> But this becomes more complicated when dealing with multiple (user_id, device_id) combinations, as the AS API would like to do.

I'm not sure I entirely follow this. As long as you `ORDER BY stream_id`, you will know that you have complete results for all but the last returned `stream_id` - so you just need to drop the last, potentially incomplete, `stream_id` from the results before returning them to the caller.

Reply:
What will the caller than call the method with next time though? If they try again with the last returned `stream_id`, at best they'll get some duplicated messages, and at worst if the number of messages belonging to a single stream ID > the limit, then we'll get in a never-ending loop. `stream_id - 1` has the same problem, and `stream_id + 1` will cause messages to be missed.

An alternative approach:

The caller provides `from` and `to` tokens, as well as an offset value. The storage function returns the latest processed `stream_id` as well as an offset, which the caller then hands right back on the next call, and so on.

This way, if we end up hitting our limit in the middle of a `stream_id`s group of messages, we can identify how much of that group was returned on the last call. We then just offset through that same group on the next call.

We'd need to ensure a `stream_id`s group of messages are always consistently ordered (`ORDER BY stream_id, device_id`). We'd also have to store this offset in the database, though that's just another column.

I think that should work, though there's likely a simpler way to do it...

Reply:
Following on from your example; suppose we call `get_new_messages_for_devices` with `current_stream_id=0, limit=2`. We add one to the limit, and make the DB query, which returns the following:

```
      user_id      |                  device_id                  | stream_id |     message_json   | instance_name 
-------------------+---------------------------------------------+-----------+--------------------+---------------
 @alice:localhost  | HAKDUDVZXZ                                  |         1 | {"content":{"A"} | master
 @alice:localhost  | HAKDUDVZXZ                                  |         2 | {"content":{"B"} | master
 @bob:localhost    | TYCGLKNPVJ                                  |         2 | {"content":{"B"} | master
```

We can now tell that, at the limit (2), we would have incomplete results for `stream_id==2`. So, `get_new_messages_for_devices` *drops*all the messages with `stream_id==2`, and returns the remainder, with `stream_pos=1`.

Next time round, the caller passes the returned `stream_pos` (ie, 1) back as `current_stream_id` (noting that `current_stream_id` is *ex*clusive) and gets all the messages with stream_id=2.

(You might remember this algorithm from https://github.com/matrix-org/synapse/pull/5156/files#diff-d9219a3564a55443579532e051d6abc32f47e8f4ba44c64459cd4f9d378e22bdR90-R121).

You're right that it fails if there are more messages with a given `stream_id` for the given devices than there are at the limit. But assuming we're talking about a limit of say 100, that would mean that the AS would have to be managing at least 100 devices... how likely a scenario is this?

Reply:
> You're right that it fails if there are more messages with a given stream_id for the given devices than there are at the limit. But assuming we're talking about a limit of say 100, that would mean that the AS would have to be managing at least 100 devices... how likely a scenario is this?

Pretty likely I'm afraid - current end-to-end encryption bridge architectures require at least one device per bridged user.

Reply:
bleh. In which case, maybe we're better off with your earlier idea of 

> we could just have public methods which only accept:
> 
>     * multiple `(user_id, device_id)` pairs, no limit
>     * a single `(user_id, device_id)` pair, limit
> 
> and both call to a unified, private method.

having to store an offset into a given stream_id feels pretty icky to me.

Reply:
https://github.com/matrix-org/synapse/pull/11215/commits/822e92a62f35f52a1cefe42d5af6bed75da93c62 refactors the two previous methods into two public methods (`get_messages_for_device` and `get_messages_for_user_devices`) which both call the private method `_get_device_messages`.

A `limit` option is only allowed for `get_messages_for_device`, which takes a single user ID / device ID pair.